docs/proposals/spec-reconciliation.md

```diff
@@ -0,0 +1,130 @@
+# Spec Reconciliation Proposal
+
+## Introduction
+Add a way for broker to sync the list of specs it has with registries while handling the following corner cases:
+ 1. Temporary loss of connection with registry
+ 2. A spec removed from the registry but has provisioned instance/s
+ 3. The service catalog sends a provision request to the broker after the broker has started bootstrapping but before it has completed bootstrapping (untimely provision request)
+ 4. The service catalog sends a provision request to the broker for specs that has been marked deleted in the broker (out of sync provision request)
+
+## Problem Description
+
+The current approach of syncing the spec from registries is as follows:
+ 1. Delete all the specs currently in the datastore
+ 2. Fetch new specs
+ 3. Add all the fetched specs to the datastore
+ 
+This leads to a bunch of bugs:
+ - [Bug 1583495](https://bugzilla.redhat.com/show_bug.cgi?id=1583495)
+ - [Bug 1577810](https://bugzilla.redhat.com/show_bug.cgi?id=1583495) 
+ - [Bug 1586345](https://bugzilla.redhat.com/show_bug.cgi?id=1586345)
+
+The underlying problem with all those bugs is that the broker is deleting specs even if it is provisioned. When the catalog tries to deprovision, bind, unbind with these dangling instances (instances with no specs) it will fail. Hence the broker needs a way to safely delete the specs.
+
+### Approaches to safely deleting specs
+
+The broker can take the following actions to safely delete the specs:
+#### Approach#1
```

Comment:
I like Approach 1 best but I'd like to change it a bit.
1. Fetch the specs from a registry
2. Fetch the specs from the datastore
3. Delete any specs in the datastore currently marked for deletion, assuming there is no instance that has a reference to this spec which is marked for deletion.
   i. if there is a reference, leave the spec alone
  ii. if there is no reference, delete the spec
4. For each spec from the registry, get the spec from the datastore
   i. if found, update it, and save to datastore
   ii. if not found, add it to the datastore
5. For each spec in the datastore, NOT in registry list, mark it for deletion.

Connection failure
   i. if connection fails, retry 10 times backing of 10 seconds each time, then give up until the next bootstrap loop runs
   if the bootstrap loop is < our retry time, I'd abort and wait for the next bootstrap loop to run.

Catalog payload should change such that it returns all specs that are NOT marked for deletion

I would **not** do reference counting, I would simply look for any instances that have a reference to a spec that was marked for deletion. If so, then we do not delete it.

Reply:
I agree with @shawn-hurley, we already frequently hit 504s with the broker route when bootstrapping I would prefer to not add too much computation to the bootstrap loop.